### ISSUE TITLE
Mode 2 gives the same result as mode 3

### ISSUE THREAD

**Issue by shivivats** [itu-p1203/itu-p1203#39]
Hi, testing with v1.10.0 on Ubuntu 22.04 installed using WSL.

Running the `p1203-standalone` command with `-m 2` gives the same result as mode 3, and even shows `"mode": 3` in the output. Have tested this with multiple vidoes, same thing. Running it with `-m 3` gives the exact same output.

Will look into the code when I find the time and get to the root of this, but wanted to raise an issue if someone had an idea on how to fix this (or if this is expected behaviour in some cases?).

Best,
Shivi

**Comment by slhck** [comment]
The extractor, as implemented, parses the entire bitstream and therefore forwards the same information to the Mode 2 evaluation as in Mode 3:

https://github.com/itu-p1203/itu-p1203/blob/38c5d55c6250b3ab6bcc5ccef19085a6055ba9b7/itu_p1203/extractor.py#L689-L693

The video model code therefore assumes that if `qpValues` are present in the input, the calculation mode is 3:

https://github.com/itu-p1203/itu-p1203/blob/38c5d55c6250b3ab6bcc5ccef19085a6055ba9b7/itu_p1203/p1203Pv.py#L595-L596

Since the `Pv` class doesn't know whether Mode 2 or 3 are intended, it simply picks Mode 3.

To fix this, we would have to:

- Make the Mode 2 extractor keep only 2% of the information
- Explicitly pass the mode to the class to correctly round-trip the information

**Comment by slhck** [comment]
I think this can be done quite easily, although I have to check whether adding an additional `mode` kwarg for the `Pv` class doesn't break some downstream dependencies (where, e.g., a subclass of `Pv` is used with modified coefficients etc.).

**Comment by shivivats** [comment]
Thanks for the quick comment! I can work on this and submit a PR when done.

Just to clarify, we're aiming for using only 2% of the `qpValues` obtained from `get_video_frame_info_ffmpeg_debug_qp`. I imagine we're gonna have to do some rounding (and thus end up using more of the information) if the array of qp values does not have elements in the multiple of 50.

I'll go ahead with this logic, but please let me know if I'm understanding it wrong.

**Comment by slhck** [comment]
> Thanks for the quick comment! I can work on this and submit a PR when done.

That would be great!

> Just to clarify, we're aiming for using only 2% of the `qpValues` obtained from `get_video_frame_info_ffmpeg_debug_qp`. I imagine we're gonna have to do some rounding (and thus end up using more of the information) if the array of qp values does not have elements in the multiple of 50.
> 
> I'll go ahead with this logic, but please let me know if I'm understanding it wrong.

Mode 2 is actually a bit more complex. The core idea of mode 2 was to only use 2% of the bitstream *itself*, not 2% of the actual QP values (i.e. 2 QP values if a whole frame had 100 macroblocks). Models would only be allowed to peek into the bitstream until they consumed 2% of its bytes, using whatever information they wanted over a sliding window of 20 seconds duration.

That is, obviously, just not practical — it was an arbitrary limitation imposed on the models to ensure a fair competition, so that no model proponent could read too much information from the stream. It has never been implemented like this in practice.

So, using only a few QP values from the entire set is a reasonable approximation for this, but as you can see, the actual _model_ for Mode 2 and 3 is the same. So the question is more about which QP values you choose (and it's very likely not the top left of every frame that will be the most meaningful 😉).

**Comment by shivivats** [comment]
I see, it is indeed a bit more complex than I thought 😄 

I'll wrack my brain and consult some people at my Uni that might know better about which QP values to use/how to get them.

Side note: was this the reason why you wrote the "hack" `ffmpeg_debug_qp` tool? To be able to use QP values instead of the bitstream since using the latter directly is not reasonable?

**Comment by slhck** [comment]
> I see, it is indeed a bit more complex than I thought 😄
> 
> I'll wrack my brain and consult some people at my Uni that might know better about which QP values to use/how to get them.

Yeah … you have a separate research project there (which QP values are the most relevant for the overall quality estimation in a limited resource scenario?).

> 
> Side note: was this the reason why you wrote the "hack" `ffmpeg_debug_qp` tool? To be able to use QP values instead of the bitstream since using the latter directly is not reasonable?

For the ITU competition we had a much more complex bitstream parser, which eventually was published here: https://github.com/Telecommunication-Telemedia-Assessment/bitstream_mode3_videoparser
It was also consequently used for the P.1204.3 bitstream model, which uses more than just the QP information.

Since for P.1203.1 Mode 3, QP is enough, the ffmpeg tool was a good compromise to not having to use the full bitstream parser (which at the time could not be made public anyway). That being said, because ffmpeg does not readily export these QP values anymore, I am currently writing a new parser that will be open sourced to the research community at some point.
